developers/contributing/designs/in-repo-test-config.md

```diff
@@ -0,0 +1,44 @@
+
+On the Zuul single job template will be defined that contains the necessary builders and publishers for executing the testing script and pubishing the test output and logs.  This job template can be used to add jobs to new projects' pipelines. This will help standardize project configuration and streamline the onboarding of new projects into BonnyCI.
+
+### In-repo testing script
+
+#### Location
+
+Projects will create a ``.bonnyci`` subdirectory at the root of the repository tree.  This directory may contain anything useful for executing tests (ie, shell libraries or helper scripts), but if tests are to be run it will need to contain an executable file ``run.sh`` that serves as an entry point to the test run.  This script is user-defined and may call out to any other scripts stored in the repo, or to binaries and scripts available elsewhere on the test slave (ie, virtualenv or pip).
+
+#### Interface
+
+The ``run.sh`` script is free to define any steps necessary to run the projects' tests.  It is required that its permissions allow it to be executed and it contains the required ``#!/bin/bash`` shebang.  Additionally, the script's return code is interpretted by BonnyCI and should be one of either:
+
+* 0 - This signifies the test run executed successfully and without errors and that th test job has passed.
+* non-zero - This signifies a test failure or error and that the test job has failed.
+
+The ``run.sh`` script will be used to invoke tests across both gate and check pipelines, and possibly other pipelines in the future.  Users may wish to differentiate their test runs between pipelines.  For example, a project may only run linting and unit tests during the initial check, but additionally run functional or integration test as part of gating.  To allow this flexibility, the pipeline in which the test is being run will be exported to the script via environment variable.  The ``BONNYCI_TEST_PIPELINE`` variable will be set to the pipeline currently running, currently either ``check`` or ``gate``.  Note that these pipeline names are user-facing and distinct from the names of the internal pipelines configured in Zuul's configuration.
+
+Regardless of a jobs passing or failing status, BonnyCI will publish console output and, optionally, any test logs made available by the project's test suite.  User's are responsible for ensuring their ``run.sh`` places any log files in the correct location on the file system for post-run archiving.  The location of this directory will be made available to the script via a ``BONNYCI_TEST_LOG_DIR`` environment variable and it can be assumed that this directory exists and is writable by the user executing tests.
+
+### Zuul Configuration
+
+Our Zuul job configuration will contain a single bonnyci job template that defines:
+
+* A publisher that publishes console logs and test log files
```

Comment:
The script will be notified of the correct location to stash logs via the BONNYCI_TEST_LOGS env var.  Whatever we end up setting that to will be hard-coded into the job definition's publisher